File: src/solutions/0200-0299/demo_206.ts
//
// function reverseList(head: ListNode | null): ListNode | null {
//     let pre: ListNode | null = null;
//     let cur: ListNode | null = head;
//
//     while (cur != null) {
//         const temp: ListNode | null = cur.next;
//         cur.next = pre;
//         pre = cur;
//         cur = temp;
//     }
//     return pre;
// };





function reverseList(head: ListNode | null): ListNode | null {
    if (head === null || head.next === null) {
        return head;
    }
    const newHead = reverseList(head.next);
    head.next.next = head;
    head.next = null;
    return newHead;
};

File: src/solutions/0200-0299/demo_206.java
//public class demo_206 {
//}
//
//
//
///*反转链表*/
//
// public class ListNode {
//      int val;
//      ListNode next;
//      ListNode() {}
//      ListNode(int val) { this.val = val; }
//      ListNode(int val, ListNode next) { this.val = val; this.next = next; }
//  }
//
//class Solution {
//    public ListNode reverseList(ListNode head) {
//        ListNode prev = null;
//        ListNode cur = head;
//        ListNode temp = null;
//        while (cur != null) {
//            temp = cur.next;// 保存下一个节点
//            cur.next = prev;
//            prev = cur;
//            cur = temp;
//        }
//        return prev;
//    }
//}